crates/sui-storage/src/indexes.rs

```diff
@@ -879,7 +884,42 @@ impl IndexStore {
             .iter()
             // The object id 0 is the smallest possible
             .skip_to(&(owner, starting_object_id))?
-            .take_while(move |((object_owner, _), _)| (object_owner == &owner))
+            .filter(move |((object_owner, _), obj_info)| {
+                // println!("filter_obj_info {:?}", obj_info);
+                let to_include: bool = match &filter {
+                    Some(SuiObjectDataFilter::StructType(struct_tag)) => {
+                        let obj_tag: StructTag = match obj_info.type_.clone().try_into() {
+                            Ok(tag) => tag,
+                            Err(_) => {
+                                return false;
+                            }
+                        };
+                        // If people do not provide type_params, we will match all type_params
+                        // e.g. `0x2::coin::Coin` can match `0x2::coin::Coin<0x2::sui::SUI>`
+                        if !struct_tag.type_params.is_empty()
+                            && struct_tag.type_params != obj_tag.type_params
+                        {
+                            return false;
+                        }
+                        return obj_tag.address == struct_tag.address
+                            && obj_tag.module == struct_tag.module
+                            && obj_tag.name == struct_tag.name;
+                    }
+                    Some(SuiObjectDataFilter::MoveModule { package, module }) => {
+                        if let ObjectType::Struct(MoveObjectType::Other(ref tag)) = obj_info.type_ {
+                            tag.address.to_string() == package.to_string()
+                                && tag.module.to_string() == module.to_string()
+                        } else {
+                            false
+                        }
```

Comment:
```suggestion
                        if let ObjectType::Struct(o) = obj_info.type_ {
                            p.address().to_string() == package.to_string()
                                && tag.module == module
                        } else {
                            false
                        }
```

We also need to support MoveObjectType::GasCoin and MoveObjectType::StakedSui

Reply:
Oh good catch!